src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_service.clj

```diff
@@ -2,27 +2,44 @@
   (:require [clojure.tools.logging :as log]
             [puppetlabs.i18n.core :refer [trs]]
             [puppetlabs.trapperkeeper.services :as tk]
-            [puppetlabs.trapperkeeper.services.protocols.filesystem-watch-service  :refer [FilesystemWatchService create-watcher]]
+            [puppetlabs.trapperkeeper.services.protocols.filesystem-watch-service :refer [FilesystemWatchService create-watcher]]
             [puppetlabs.trapperkeeper.services.watcher.filesystem-watch-core :as watch-core])
-  (:import (java.io IOException)
-           (java.nio.file WatchService)))
+  (:import (java.nio.file WatchService)))
 
 (tk/defservice filesystem-watch-service
   FilesystemWatchService
   [[:ShutdownService shutdown-on-error]]
 
   (init
     [this context]
-    (assoc context :watchers (atom [])))
+    (assoc context :watchers (atom [])
+                   :watchers-futures (atom [])
+                   :stopping? (atom false)))
 
   (stop
     [this context]
+    (log/info (trs "Shutting down watcher service"))
+    (reset! (:stopping? context) true)
     ;; Shut down the WatchServices
     (doseq [watcher @(:watchers context)]
       (try
         (.close ^WatchService (:watch-service watcher))
-        (catch IOException e
+        (catch Throwable e
           (log/warn e (trs "Exception while closing watch service")))))
+
+    (doseq [watchers-future @(:watchers-futures context)]
+      (try
+        (future-cancel watchers-future)
+        (loop [tries 100]
+          (if (and (pos? tries) (not (future-done? watchers-future)))
+            (do
+              (Thread/sleep 10)
+              (recur (dec tries)))
+            (log/debug (trs "Future completed with {0} tries left" tries))))
```

Comment:
I think it would be better to say "Future completed after X tries" rather than asking folks know that it will try 100 times.